klab.hub/src/main/java/org/integratedmodelling/klab/hub/tasks/controllers/AgreementController.java

```diff
@@ -0,0 +1,89 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class AgreementController {
+
+    @Autowired
+    UserAgreementService service;
+
+    @GetMapping(value = API.HUB.USER_AGREEMENTS)
+    @RolesAllowed({"ROLE_ADMINISTRATOR", "ROLE_SYSTEM"})
+    public ResponseEntity< ? > revokeAgreementOfUser(
+            @PathVariable("id") String username) {
+        Collection<Agreement> agreements;
+        try {
+            agreements = service.getAgreementsFromUser(username);
+        } catch (BadRequestException e) {
+            return ResponseEntity
+                    .status(HttpStatus.BAD_REQUEST)
+                    .body(e.getMessage());
+        }
+
+        return ResponseEntity.status(HttpStatus.ACCEPTED)
+                .body(agreements);
+    }
+
+    @DeleteMapping(value = API.HUB.USER_AGREEMENT_ID)
+    @RolesAllowed({"ROLE_ADMINISTRATOR", "ROLE_SYSTEM"})
+    public ResponseEntity< ? > revokeAgreementOfUser(
+            @PathVariable("id") String username,
+            @PathVariable("agreement-id") String agreementId) {
+        try {
+            service.revokeAgreementFromUser(username, agreementId);
+        } catch (BadRequestException e) {
+            return ResponseEntity
+                    .status(HttpStatus.BAD_REQUEST)
+                    .body(e.getMessage());
+        }
+
+        return ResponseEntity.status(HttpStatus.ACCEPTED)
+                .body(String.format("Agreement %s of user %s has been revoked.", agreementId, username));
+    }
+
+    private boolean validateAgreementId(Agreement agreement, String agreementId) {
+        if (agreement.getId() == null) {
+            agreement.setId(agreementId);
+            return true;
+        }
+        return agreement.getId() == agreementId;
+    }
+    
+    @PatchMapping(value = API.HUB.USER_AGREEMENT_ID, consumes = "application/json")
+    @RolesAllowed({"ROLE_ADMINISTRATOR", "ROLE_SYSTEM"})
+    public ResponseEntity< ? > patchAgreementOfUser(
+            @PathVariable("id") String username,
+            @PathVariable("agreement-id") String agreementId,
+            @RequestBody Agreement agreement) {
+        if(!validateAgreementId(agreement, agreementId)) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                    .body("Cannot modify agreement. Unclear id reference.");
+        }
+        try {
+            service.patchAgreementOfUser(username, agreement);
```

Comment:
The _patch_ idea is based on the the [HTTP verb](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH) (HTTP layer) right? What we're doing at the service layer is _updating_ the agreement. I would suggest to not mix layer concepts/names.